### PR TITLE
[ETCM-841] Rework protocol negotiation

### PR DESCRIPTION
**Pull request by dzajkowski**
Mantis implements a protocol version ETC/64 which is conflicting with ETH/64.
This PRs goal is to enable the coexistence of both and some future proofing (renames, simple protocol negotiation).

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/network/handshaker/EtcHelloExchangeState.scala [hunk 1]
```diff
@@ -26,19 +26,15 @@ case class EtcHelloExchangeState(handshakerConfiguration: EtcHandshakerConfigura
...
-      log.debug(
-        s"Connected peer does not support eth ${ProtocolVersions.PV63.toByte} / ${ProtocolVersions.PV64.toByte} protocol. Disconnecting."
-      )
-      DisconnectedState(Disconnect.Reasons.IncompatibleP2pProtocolVersion)
+    capability match {
+      case ProtocolVersions.ETC64 => EtcNodeStatus64ExchangeState(handshakerConfiguration)
+      case ProtocolVersions.ETH63 => EtcNodeStatus63ExchangeState(handshakerConfiguration)
+      case _ =>
+        log.debug(
+          s"Connected peer does not support ${ProtocolVersions.ETH63} / ${ProtocolVersions.ETC64} protocol. Disconnecting."
```
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/network/handshaker/EtcHelloExchangeState.scala]: shall we align the log msg format?
```suggestion
          "Connected peer does not support {} / {} protocol. Disconnecting.", ProtocolVersions.ETH63, ProtocolVersions.ETC64
```

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala [hunk 2]
```diff
@@ -7,6 +7,22 @@ import io.iohk.ethereum.rlp.{RLPEncodeable, RLPException, RLPList, RLPSerializab
 case class Capability(name: String, version: Byte)
 
 object Capability {
+  def negotiate(c1: List[Capability], c2: List[Capability]): Option[Capability] = {
+    c1.intersect(c2).maxByOption(_.version) // FIXME ignores branches and other protocols
+  }
```
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala]: curly braces can be omitted here

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala [hunk 3]
```diff
@@ -7,6 +7,22 @@ import io.iohk.ethereum.rlp.{RLPEncodeable, RLPException, RLPList, RLPSerializab
...
+
+  private val pattern = "(.*)/(\\d*)".r
+
+  def from(protocolVersion: String): Capability =
+    protocolVersion match {
+      case pattern(name, version) => Capability(name, version.toByte)
+      case _ => throw new RuntimeException(s"Unable to parse capability $protocolVersion")
```
Comment (AnastasiiaL) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala:26]:
```suggestion
      case _ => throw new RuntimeException("Unable to parse capability {}", protocolVersion)
```

Comment (dzajkowski) [reply src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala:26]: I'm not convinced that RuntimeException accepts template strings.

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala [hunk 4]
```diff
@@ -7,6 +7,21 @@ import io.iohk.ethereum.rlp.{RLPEncodeable, RLPException, RLPList, RLPSerializab
 case class Capability(name: String, version: Byte)
 
 object Capability {
+  def negotiate(c1: List[Capability], c2: List[Capability]): Option[Capability] =
+    c1.intersect(c2).maxByOption(_.version) // FIXME ignores branches and other protocols
+
+  private val pattern = "(.*)/(\\d*)".r
+
+  def from(protocolVersion: String): Capability =
```
Comment (AurelienRichez) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala]: Could we make it clear that this function throws ? `parseUnsafe` maybe ? 
Or make it return an option and let `BlockchainConfig` throw in case of none

File: src/main/scala/io/iohk/ethereum/nodebuilder/NodeBuilder.scala [hunk 5]
```diff
@@ -246,7 +248,7 @@ trait PeerManagerActorBuilder {
       EthereumMessageDecoder,
       discoveryConfig,
       blacklist,
-      Config.Network.protocolVersion
+      blockchainConfig.capabilities // TODO replace with a list of capabilities
```
Comment (AurelienRichez) [inline src/main/scala/io/iohk/ethereum/nodebuilder/NodeBuilder.scala]: Isn't it already a list of capabilities ?

File: src/main/scala/io/iohk/ethereum/utils/BlockchainConfig.scala [hunk 6]
```diff
@@ -122,6 +124,8 @@ object BlockchainConfig {
     val allowedMinersPublicKeys = readPubKeySet(blockchainConfig, "allowed-miners")
 
     val ecip1099BlockNumber: BigInt = BigInt(blockchainConfig.getString("ecip1099-block-number"))
+    val capabilities: List[Capability] =
+      blockchainConfig.getStringList("capabilities").asScala.toList.map(Capability.from)
```
Comment (AurelienRichez) [inline src/main/scala/io/iohk/ethereum/utils/BlockchainConfig.scala]: I guess we should also check that the capability in the config is indeed a capability implemented by Mantis ?

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala [hunk 7]
```diff
@@ -7,6 +7,21 @@ import io.iohk.ethereum.rlp.{RLPEncodeable, RLPException, RLPList, RLPSerializab
 case class Capability(name: String, version: Byte)
 
 object Capability {
+  def negotiate(c1: List[Capability], c2: List[Capability]): Option[Capability] =
+    c1.intersect(c2).maxByOption(_.version) // FIXME ignores branches and other protocols
```
Comment (AurelienRichez) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala]: It seems that it duplicates the `best` function here.
We should probably have a single point where the best protocol is chosen, no ?
Something like :
```suggestion
   // FIXME ignores branches and other protocols
   c1.intersect(c2) match {
     case Nil => None
     case nonEmpty => best(nonEmpty)
   }
```

**Comment by leo-bogastry** [comment]
The ETCM-841 ticket description mentions "The task is to rename the whole protocol to PV164 and mark @depracated". Are you thinking keeping it as ETC/64 instead? With or without deprecation?

**Comment by dzajkowski** [comment]
@LeonorLunatech As I mentioned during refinement - I went through a few iterations on how to deal with this issue and by explicitly stating the namespace (prefix) mantis can have any number of protocols. The only limitation will be the negotiation mechanism.

TL;DR the deprecation is not needed with the prefix. This way mantis stays backwards compatible.

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala (hunk 4 above)
Comment (dzajkowski) [reply src/main/scala/io/iohk/ethereum/network/p2p/messages/Capability.scala]: good point. I'll stick to the exception. not the best solution but we need mantis to fail when configs are not properly defined.

**Review by AurelienRichez — APPROVED** [review]

**Review by leo-bogastry — APPROVED** [review]
